lmdeploy/pytorch/kernels/cuda/fused_lora.py

```diff
@@ -0,0 +1,188 @@
+
+    if pid_m * BLOCK_SIZE_M >= M:
+        return
+
+    offs_m = (seq_start + pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M))
+    offs_n = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N))
+
+    mask_cm = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M) < M
+    if rank == 0:
+        offs_cm = offs_m
+        offs_cn = offs_n
+        c_ptrs = c_ptr + stride_cm * offs_cm[:, None] + stride_cn * offs_cn[
+            None, :]
+        c_mask = mask_cm[:, None] & (offs_cn[None, :] < N)
+        tl.store(c_ptrs, 0, mask=c_mask)
+        return
+
+    offs_am = (seq_start +
+               (pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)) % M)
+    offs_r = rank_start + tl.arange(0, BLOCK_SIZE_R) % rank
+    offs_k = tl.arange(0, BLOCK_SIZE_K)
+    a_ptrs = a_ptr + (offs_am[:, None] * stride_am +
+                      offs_k[None, :] * stride_ak)
+    la_ptrs = lora_a_ptr + (offs_k[:, None] * stride_lak +
+                            offs_r[None, :] * stride_lar)
+
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_R), dtype=tl.float32)
+    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
+        # Load the next block of A and B
+        # If it is out of bounds, set it to 0.
+        a = tl.load(a_ptrs,
+                    mask=offs_k[None, :] < K - k * BLOCK_SIZE_K,
+                    other=0.0)
+        la = tl.load(la_ptrs,
+                     mask=offs_k[:, None] < K - k * BLOCK_SIZE_K,
+                     other=0.0)
+        # We accumulate along the K dimension.
+        accumulator += tl.dot(a, la)
+        # Advance the ptrs to the next K block.
+        a_ptrs += BLOCK_SIZE_K * stride_ak
+        la_ptrs += BLOCK_SIZE_K * stride_lak
+    ar = accumulator.to(lora_b_ptr.dtype.element_ty)
+
+    offs_lbn = offs_n % N
+    lb_ptrs = lora_b_ptr + (offs_r[:, None] * stride_lbr +
+                            offs_lbn * stride_lbn)
+    lb = tl.load(lb_ptrs, mask=tl.arange(0, BLOCK_SIZE_R)[:, None] < rank)
+
+    c = tl.dot(ar, lb)
+
+    scaling = tl.load(scaling_ptr + adapter_id)
+    c *= scaling
+
+    c = c.to(c_ptr.dtype.element_ty)
+    offs_cm = offs_m
+    offs_cn = offs_n
+    c_ptrs = c_ptr + stride_cm * offs_cm[:,
+                                         None] + stride_cn * offs_cn[None, :]
+    c_mask = mask_cm[:, None] & (offs_cn[None, :] < N)
+    tl.store(c_ptrs, c, mask=c_mask)
+
+
+def fused_lora(input: torch.Tensor, lora_a: torch.Tensor, lora_b: torch.Tensor,
+               scaling: torch.LongTensor, rank_start: torch.LongTensor,
+               ranks: torch.LongTensor, seq_start: torch.LongTensor,
+               seq_lens: torch.LongTensor, adapter_ids: torch.LongTensor,
+               max_rank: int, max_seqlen: int):
+    """fused lora."""
+
+    def grid(META):
+        ret = ((triton.cdiv(max_seqlen, META['BLOCK_SIZE_M']) *
+                triton.cdiv(N, META['BLOCK_SIZE_N'])), batch_size)
+        return ret
+
+    assert input.dim() == 2
+    batch_size = seq_lens.numel()
```

Comment:
Here, we know how many sessions are running. Why do we use `seq_lens.numel()`?

Reply:
cudagraph support require static grid size.